src/main/java/org/opensearch/agent/tools/DataDistributionTool.java

```diff
@@ -0,0 +1,1298 @@
+            double upperBound = i == numGroups - 1 ? max : min + (i + 1) * groupSize;
+            String groupLabel = String.format(Locale.ROOT, "%.1f-%.1f", lowerBound, upperBound);
+
+            groupedSelectionDist.put(groupLabel, 0.0);
+            groupedBaselineDist.put(groupLabel, 0.0);
+
+            for (double numKey : numericKeys) {
+                boolean belongs = i == numGroups - 1
+                    ? numKey >= lowerBound && numKey <= upperBound
+                    : numKey >= lowerBound && numKey < upperBound;
+                if (belongs) {
+                    String strKey = String.valueOf(numKey);
+                    groupedSelectionDist.merge(groupLabel, selectionDist.getOrDefault(strKey, 0.0), Double::sum);
+                    groupedBaselineDist.merge(groupLabel, baselineDist.getOrDefault(strKey, 0.0), Double::sum);
+                }
+            }
+        }
+
+        return new GroupedDistributions(groupedSelectionDist, groupedBaselineDist);
+    }
+
+    /**
+     * Formats field analysis results into summary data items
+     *
+     * @param differences List of field analysis results
+     * @param maxResults Maximum number of results to return
+     * @return Formatted list of summary data items
+     */
+    private List<SummaryDataItem> formatComparisonSummary(List<FieldAnalysis> differences, int maxResults) {
+        return differences.stream().filter(diff -> diff.divergence > 0).limit(maxResults).map(diff -> {
+            Set<String> allKeys = new HashSet<>(diff.selectionDist.keySet());
+            allKeys.addAll(diff.baselineDist.keySet());
+
+            List<ChangeItem> changes = allKeys.stream().map(value -> {
+                double selectionPercentage = Math.round(diff.selectionDist.getOrDefault(value, 0.0) * 100.0) / 100.0;
+                double baselinePercentage = Math.round(diff.baselineDist.getOrDefault(value, 0.0) * 100.0) / 100.0;
+                return new ChangeItem(value, selectionPercentage, baselinePercentage);
+            }).collect(Collectors.toList());
+
+            List<ChangeItem> topChanges = changes
+                .stream()
+                .sorted(
+                    (a, b) -> Double
+                        .compare(
+                            Math.max(b.baselinePercentage, b.selectionPercentage),
+                            Math.max(a.baselinePercentage, a.selectionPercentage)
+                        )
+                )
+                .limit(10)
+                .collect(Collectors.toList());
+
+            return new SummaryDataItem(diff.field, diff.divergence, topChanges);
+        }).collect(Collectors.toList());
+    }
+
+    /**
+     * Builds query conditions from filter map for DSL queries
+     *
+     * @param filterMap Filter conditions as map
+     * @param queryBuilder Query builder to add conditions to
+     */
+    private void buildQueryFromMap(Map<String, Object> filterMap, BoolQueryBuilder queryBuilder) {
```

Comment:
This method has too deep nested loops, let's optimize it by splitting into several small methods.